File: cpp/chapter_stack_and_queue/linkedlist_stack.cpp
/**
 * File: linkedlist_stack.cpp
 * Created Time: 2022-11-28
 * Author: qualifier1024 (2539244001@qq.com)
 */

#include "../utils/common.hpp"
class LinkedListStack{
    private:
        ListNode *stackTop;
        int stkSize;
    public:
        LinkedListStack(){
            stackTop=NULL;
            stkSize=0;
        }
        
        ~LinkedListStack(){
            freeMemoryLinkedList(stackTop);    
        }
        //获取栈的长度
        int size(){
            return stkSize;   
        }
        //判断栈是否为空
        bool isEmpty(){
            return stkSize==0;   
        }
        //入栈
        void push(int val){
            ListNode *newNode=new ListNode(val);
            newNode->next=stackTop;
            stackTop=newNode;
            stkSize++;
        }
        //出栈
        int pop(){
            if(isEmpty()){
                cout<<"栈为空，无法出栈"<<endl;
                return -1;
            }
            int num=top();
            ListNode *tmp=stackTop;
            stackTop=stackTop->next;
            stkSize--;
            int val=tmp->val;
            delete tmp;
            return val;
        }
        //获取栈顶元素
        int top(){
            if(isEmpty()){
                cout<<"栈为空，无法获取栈顶元素"<<endl;
                return -1;
            }
            return stackTop->val;
        }
        //将List 转化为 Array并返回
        vector<int> toArray(){
            vector<int> res(size());
            ListNode *node=stackTop;
            for(int i=size()-1;i>=0;i--){
                res[i]=node->val;
                node=node->next;
            }
            return res;
        }
        
};

int main(){
    return 0;
}

File: python/chapter_stack_and_queue/linkedlist_stack.py
import sys
from pathlib import Path

sys.path.append(str(Path(__file__).parent.parent))
from modules import ListNode

class LinkedListStack:
    # 初始化栈
    def _init_(self):
        self._peek :ListNode = None 
        self.size = 0
    # 长度
    def size(self)->int:
        return self.size
    # 判空
    def is_empty(self)->bool:
        return self.size==0
    # 入栈
    def push(self,val:int):
        node = ListNode(val)
        node.next = self._peek
        self._peek = node
        self.size+=1
    # 出栈
    def pop(self)->int:
        if self.is_empty():
            raise Exception("Stack is empty")
        node = self._peek
        self._peek = node.next
        node.next = None
        self.size-=1
        return node.val
    # 获取栈顶元素
    def peek(self)->int:
        if self.is_empty():
            raise Exception("Stack is empty")
        return self._peek.val
    # 打印
    def to_list(self)->list[int]:
        res = []
        cur = self._peek
        while cur:
            res.append(cur.val)
            cur = cur.next
        res.reverse()
        return res
    